### DISCUSSION TITLE
What does (ZydisFormatterBufferAppend do ?

### DISCUSSION THREAD

**Discussion by 440bx** [zyantific/zydis#494] [Q&A] [answered]
Hello, 

In the example program Formatter01, in the function ZydisFormatterPrintAddressAbsolute there is the following statement (line 89 or thereabouts):

`ZYAN_CHECK(ZydisFormatterBufferAppend(buffer, ZYDIS_TOKEN_SYMBOL));`

after reading the documentation, I can't figure out what that statement does and why it's there.  I commented it out and the program's behavior and output was unchanged.

The questions are: what's that statement for ? what does it do ? and what are the consequences of not calling the function ?

Thank you for your help.

**Comment by flobernd** [comment, marked as answer]
https://github.com/zyantific/zydis/blob/ffde0f46398a86417c860462f6af0556331cb5bb/examples/Formatter01.c#L88-L91

This basically marks the following string as a "symbol". There are different types of tokens: https://github.com/zyantific/zydis/blob/master/include/Zydis/FormatterBuffer.h#L55

When using the `ZydisFormatterFormat` API, this does not make a difference. If you are using `ZydisFormatterTokenize` instead, you receive a list of [ZydisFormatterToken](https://github.com/zyantific/zydis/blob/master/include/Zydis/FormatterBuffer.h#L138) in which each token has a type and a value (that can be retrieved using [ZydisFormatterTokenGetValue](https://github.com/zyantific/zydis/blob/master/include/Zydis/FormatterBuffer.h#L204)).

That functionality is e.g. used in `ZydisInfo` for color coding the individual parts of the instruction:
https://github.com/zyantific/zydis/blob/master/tools/ZydisToolsShared.c#L158

**Reply by 440bx** [reply]
Wonderful !!   Thank you Florian, now I understand what that's for.    If I understood what you said correctly, it is a way of informing the formatter the type of token that is being added.  Knowing the token type may be needed to take some actions, e.g, coloring as done in ZydisInfo.

Thanks again!
